libclamav_rust/src/codesign.rs

```diff
@@ -0,0 +1,588 @@
+                #[allow(unused_unsafe)]
+                match unsafe { CStr::from_ptr(path_str) }.to_str() {
+                    Err(e) => {
+                        warn!("Intermediate path string is not valid unicode: {}", e);
+                        return None;
+                    }
+                    Ok(s) => Some(s),
+                }
+            };
+
+            if let Some(path_str) = path_str {
+                match Path::new(path_str).canonicalize() {
+                    Ok(path) => Some(path),
+                    Err(e) => {
+                        warn!(
+                            "Invalid intermediate certificate path: '{}' {}",
+                            path_str, e
+                        );
+                        None
+                    }
+                }
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    let signing_key_path_str = validate_str_param!(signing_key_path_str);
+    let signing_key_path = match Path::new(signing_key_path_str).canonicalize() {
+        Ok(p) => p,
+        Err(e) => {
+            return ffi_error!(
+                err = err,
+                Error::SignFailed(format!(
+                    "Invalid signing key path '{}': {}",
+                    signing_key_path_str, e
+                ))
+            );
+        }
+    };
+
+    match sign_file(
+        &target_file_path,
+        signature_file_path,
+        &signing_cert_path,
+        &intermediate_cert_paths,
+        &signing_key_path,
+        append,
+    ) {
+        Ok(()) => {
+            debug!("File signed successfully");
+            true
+        }
+        Err(e) => {
+            ffi_error!(err = err, e)
+        }
+    }
+}
+
+/// Signs a file.
+/// The signature is appended to the signature file.
+pub fn sign_file<P>(
+    target_file_path: &Path,
+    signature_file_path: &Path,
+    signing_cert_path: &Path,
+    intermediate_cert_paths: &[P],
+    signing_key_path: &Path,
+    append: bool,
+) -> Result<(), Error>
+where
+    P: AsRef<Path>,
+{
+    let signer = Signer::new(signing_key_path, signing_cert_path, intermediate_cert_paths)?;
+
+    let data = std::fs::read(target_file_path)?;
+    let pkcs7 = signer.sign(&data)?;
+
+    // Now convert the pkcs7 to a DigitalSig struct which may be converted to a .sign file signature line.
+    let signature = DigitalSig::Pkcs7(pkcs7);
+    let mut sig_bytes: SigBytes = SigBytes::new();
+    signature.append_sigbytes(&mut sig_bytes)?;
+
+    let mut writer = {
+        let mut options = std::fs::OpenOptions::new();
+        options.write(true).create(true);
+
+        if signature_file_path.exists() {
+            if append {
+                options.append(true);
+            } else {
+                options.truncate(true);
+            }
+        }
+
+        let mut writer = std::io::BufWriter::new(options.open(signature_file_path)?);
+
+        if !signature_file_path.exists() || !append {
```

Comment:
```suggestion
        if !signature_file_path.exists() {
```

Should still create the header in append mode.